worker/buildbot_worker/test/test_util_hangcheck.py

```diff
@@ -0,0 +1,246 @@
+        """
+        If the connection is closed, the underlying protocol is informed.
+        """
+        clock = Clock()
+        wrapped_protocol = AccumulatingProtocol()
+        protocol = HangCheckProtocol(wrapped_protocol, reactor=clock)
+        transport = StringTransport()
+
+        transport.protocol = protocol
+        protocol.makeConnection(transport)
+        reason = ConnectionDone("Bye.")
+        protocol.connectionLost(
+            Failure(reason)
+        )
+
+        self.assertTrue(wrapped_protocol.closed)
+        self.assertEqual(
+            wrapped_protocol.closedReason.value,
+            reason,
+        )
+
+    def test_disconnect_cancels_timeout(self):
+        """
+        If the connection is closed, the hang check is cancelled.
+        """
+        clock = Clock()
+        protocol = HangCheckProtocol(
+            Protocol(),
+            reactor=clock,
+        )
+        transport = StringTransport()
+
+        transport.protocol = protocol
+        protocol.makeConnection(transport)
+        protocol.connectionLost(
+            Failure(ConnectionDone("Bye."))
+        )
+
+        assert_clock_idle(self, clock)
+
+    def test_data_and_disconnect(self):
+        """
+        If the connection receives data and then is closed, no error results.
+        """
+        clock = Clock()
+        protocol = HangCheckProtocol(
+            Protocol(),
+            reactor=clock,
+        )
+        transport = StringTransport()
+
+        transport.protocol = protocol
+        protocol.makeConnection(transport)
+        protocol.dataReceived("some-data")
+        protocol.connectionLost(
+            Failure(ConnectionDone("Bye."))
+        )
+
+        assert_clock_idle(self, clock)
+
+
+def reportUnhandledErrors(case, d):
+    """
+    Make sure that any unhandled errors from the
+    given deferred are reported when the test case
+    ends.
+    """
+    def cleanup():
+        if isinstance(d.result, Failure):
+            return d
+    case.addCleanup(cleanup)
+    return d
+
+
+def listen(case, endpoint, factory):
+    """
+    Listen on an endpoint and cleanup when the
+    test case ends.
+    """
+    d = endpoint.listen(factory)
+
+    def registerCleanup(listeningPort):
+        case.addCleanup(listeningPort.stopListening)
+        return listeningPort
+    d.addCallback(registerCleanup)
+    return reportUnhandledErrors(case, d)
+
+
+class EndToEndHangCheckTests(TestCase):
+    """
+    End to end test for HangCheckProtocol.
+    """
+
+    def test_http(self):
+        """
+        When connecting to a HTTP server, a PB connection times
+        out.
+        """
+        result = Deferred()
+
+        site = Site(Data("", "text/plain"))
+        client = HangCheckFactory(
+            PBClientFactory(), lambda: result.callback(None))
+
+        self.patch(HangCheckProtocol, '_HUNG_CONNECTION_TIMEOUT', 0.1)
+
+        listen(self, TCP4ServerEndpoint(
+            reactor, 0), site).addCallback(
```

Comment:
nodeJS/lisp style. I'm lost :)

I would avoid to make a multiline lambda, and make a proper def

    server_d = listen(self, TCP4ServerEndpoint(reactor, 0), site)
    @server_d.addCallback
    def buildClient(listening_port):
        TCP4ClientEndpoint(reactor, '127.0.0.1', listening_port.getHost().port).connect(client)